Source/Utils/WarpXMovingWindow.cpp

```diff
@@ -56,25 +56,82 @@ void
 WarpX::UpdatePlasmaInjectionPosition (amrex::Real a_dt)
 {
     int dir = moving_window_dir;
-    // Continuously inject plasma in new cells (by default only on level 0)
-    if (WarpX::warpx_do_continuous_injection and (WarpX::gamma_boost > 1)){
-        // In boosted-frame simulations, the plasma has moved since the last
-        // call to this function, and injection position needs to be updated
-        current_injection_position -= WarpX::beta_boost *
+
+    // Loop over species
+    const int n_species = mypc->nSpecies();
+    for (int i=0; i<n_species; i++)
+    {
+        const WarpXParticleContainer& pc = mypc->GetParticleContainer(i);
+        const std::vector<std::string>& species_names = mypc->GetSpeciesNames();
+
+        // Continuously inject plasma in new cells (by default only on level 0)
+        if (pc.doContinuousInjection())
+        {
+            const std::unique_ptr<PlasmaInjector> plasma_injector
+                = std::make_unique<PlasmaInjector>(pc.getSpeciesId(), species_names.at(i));
+
+            // Get bulk momentum and velocity of plasma
+            amrex::XDim3 u_bulk = amrex::XDim3{0._rt, 0._rt, 0._rt};
+            amrex::Real v_bulk = 0._rt;
+            if (dir == 0)
+            {
+                // dir=0 is z in 1D, x in 2D, x in 3D
+#if defined(WARPX_DIM_1D_Z)
+                u_bulk = plasma_injector->getInjectorMomentum()->getBulkMomentum(0._rt, 0._rt, current_injection_position[i]);
+                v_bulk = PhysConst::c * u_bulk.z / std::sqrt(1._rt + u_bulk.z*u_bulk.z);
+#else // 2D, 3D
+                u_bulk = plasma_injector->getInjectorMomentum()->getBulkMomentum(current_injection_position[i], 0._rt, 0._rt);
+                v_bulk = PhysConst::c * u_bulk.x / std::sqrt(1._rt + u_bulk.x*u_bulk.x);
+#endif
+            }
+            else if (dir == 1)
+            {
+                // dir=1 is nothing in 1D, z in 2D, y in 3D
+                // (we do not expect to enter this code block in 1D)
+#if defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+                u_bulk = plasma_injector->getInjectorMomentum()->getBulkMomentum(0._rt, 0._rt, current_injection_position[i]);
```

Comment:
@WeiqunZhang
I tested the code on a local GPU and I confirm that the code crashes with segfault at this line (running in DEBUG mode), unless I set `amrex.the_arena_is_managed = 1` in the input file. Do you have ideas of why calling `plasma_injector->getInjectorMomentum()->getBulkMomentum` here would cause an issue like that?

Reply:
It looks like `getInjectorMomentum` returns the device version of the injector (`d_inj_mom`):
https://github.com/ECP-WarpX/WarpX/blob/development/Source/Initialization/PlasmaInjector.cpp#L743

It seems that we should create another function that return `h_inj_mom` instead. (Or make that attribute `public`.)

Reply:
Thanks for catching this. Do you have a preference between the two solutions?

Reply:
It's probably better to have a new function returning the host version of the injector.